Add locations to filter in Submit History

mkangia opened the pull request:
This adds capability to filter for locations and their users in SubmitHistory and CaseList. The locations were already rendered but are not filtered by.

Also added locations filtering for Reassign Cases and Manage Forms, moved out to PR #14382 

Review better at once 🏛 
http://manage.dimagi.com/default.asp?243862

mkangia committed:
Update options_url to render locations for Submit history

mkangia committed:
filter by locations selected on SubmitHistory

mkangia committed:
Make submit history location safe for access by restricted user

esoergel commented on corehq/apps/reports/filters/users.py:
```diff
@@ -176,7 +176,7 @@ class ExpandedMobileWorkerFilter(BaseMultipleOptionFilter):
     placeholder = ugettext_lazy(
         "Specify groups and users to include in the report")
     is_cacheable = False
-    options_url = 'emwf_options'
```
I think this means `Locations` will start showing up in the dropdown for all reports, right?

mkangia replied:
Yup! This does not belong here. Reverted in following commits

mkangia committed:
revert changing options_url for ExpandedMobileWorkerFilter and change the filter used instead

mkangia committed:
Handle case when no locations in filter resulting in all users getting included in filter

dimagimon commented:
There are 5 errors:

* corehq/apps/reports/filters/users.py, line 350 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 351 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 352 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 353 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 354 - E122 continuation line missing indentation or outdented

mkangia committed:
add scope of accessible forms for restricted user at SubmitHistory

mkangia committed:
filter by users at selected location ids for Case List

mkangia committed:
make case list location safe

dimagimon commented:
There are 5 errors:

* corehq/apps/reports/filters/users.py, line 350 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 351 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 352 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 353 - E122 continuation line missing indentation or outdented
* corehq/apps/reports/filters/users.py, line 354 - E122 continuation line missing indentation or outdented

mkangia committed:
fix dimagimon complaint [skip ci]

mkangia commented:
@esoergel Its great that Submit History and Case List could be made location compatible with just minor changes. Good to review.

esoergel commented on corehq/apps/locations/permissions.py:
```diff
@@ -99,7 +99,7 @@
 LOCATION_SAFE_TASTYPIE_RESOURCES = set()
 
-LOCATION_SAFE_HQ_REPORTS = set()
+LOCATION_SAFE_HQ_REPORTS = set(['case_list'])
```
I just tried it out locally, you can just add `@location_safe` to the class right at the top: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/standard/cases/basic.py#L185

mkangia replied:
Ah! Thank you Mr. Hunt. I think i was using location safe over some other class.
This works.

esoergel commented on corehq/apps/reports/filters/users.py:
```diff
@@ -342,9 +344,17 @@ def pull_users_and_groups(cls, domain, mobile_user_and_group_slugs,
+                user_ids=user_ids_at_locations_and_descendants(location_ids),
+                simplified=True,
+                CommCareUser=CommCareUser,
+            )
```
Oh nice!  This makes a lot of sense, as it will always be empty for reports that don't use locations

mkangia replied:
Also in case of reports with locations, this returned all users in case no location was selected which was weird. Will check that out later.

benrudolph replied:
just noting that we should probably respect the `limit_user_ids` variable here. in the application status report, it limits the user ids to those who have submitted to the application selected. this would ignore that and include all user ids

mkangia replied:
Thank you @benrudolph.
Made corresponding changes.

esoergel commented on corehq/apps/reports/standard/inspect.py:
```diff
@@ -92,16 +94,27 @@ def _form_filter(form):
+                 .filter(self._get_users_filter(mobile_user_and_group_slugs)))
+
+        if not self.request.can_access_all_locations:
+            query = query.filter(self.scope_filter())
```
:+1:

mkangia committed:
Use the correct way for making case list location safe

mkangia committed:
allow only web user with access to all locations view all data and all project data

mkangia committed:
add scope filter for restricted user

mkangia committed:
minor code improvement

dimagimon commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -114,42 +134,53 @@ def case_owners(self):
+            )
+
+
+            # Get group ids for each group that was specified
```
E303 too many blank lines (2)

dimagimon commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -114,42 +134,53 @@ def case_owners(self):
+            selected_sharing_group_ids = EMWF.selected_sharing_group_ids(mobile_user_and_group_slugs)
+
+            report_group_q = HQESQuery(index="groups").domain(self.domain)\
+                                               .doc_type("Group")\
```
E127 continuation line over-indented for visual indent

mkangia committed:
fix dimagimon complaints [skip ci]

esoergel commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -134,45 +134,53 @@ def case_owners(self):
-        selected_sharing_group_ids = EMWF.selected_sharing_group_ids(mobile_user_and_group_slugs)
+        # Get ids for each sharing group that contains a user from selected_reporting_group_users OR
+        # a user that was specifically selected
+        if selected_reporting_group_users and selected_user_ids:
```
This should be `or`

mkangia replied:
good catch! Thanks.

mkangia commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -70,7 +85,10 @@ def _build_query(self):
+            selected_case_owners = self.case_owners
+            if selected_case_owners:
+                query = query.owner(selected_case_owners)
+            query = query.OR(self.scope_filter())
```
oh yes.. should not be applied to restricted users. will fix that.

mkangia committed:
apply scope filter only for restricted users

mkangia committed:
fix condition for filtering by users for selected group or selected users

mkangia committed:
avoid duplicate check for accessible user ids

mkangia committed:
move filter changes to CaseReassignmentInterface from CaseListMixin to avoid conflicting with inheriting classes

mkangia committed:
Make Bulk form management interface and case reassignment views location safe

mkangia committed:
also move filter changes to CaseListReport since they were removed from CaseListMixin

mkangia committed:
render default response in case nothing selected for restricted user 

Achieved by skipping users_filter as that adds a blank list for user_ids

mkangia committed:
force user_ids search only for restricted user

mkangia committed:
extract CaseExportScope

mkangia committed:
Revert changes for Case Reassignment and Manage Forms

Extract changes for above two to a different branch instead as its getting tricky

mkangia committed:
Make case reassignment Location Safe

mkangia committed:
Restrict Case reassignment to only accessible users

mkangia committed:
make 'Manage Forms' location safe

mkangia committed:
Avoid empty result for an unrestricted user for blank filter

mkangia committed:
Make Report save view location safe for favourites for both SubmitHistory and CaseList

mkangia commented:
@esoergel extracted out changes for Case Reassignment and Manage forms sections in #14382 since those could block rolling out changes for SubmitHistory and CaseList.

esoergel commented on corehq/apps/reports/standard/inspect.py:
```diff
@@ -140,6 +155,7 @@ def total_records(self):
         return int(self.es_query_result.total)
 
 
+@location_safe
```
This all looks good to me!

esoergel commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -24,6 +26,129 @@
 from .data_sources import CaseInfo, CaseDisplay
 
 
+class CaseExportScope(object):
```
This is an awful lot of copy-pasted code - this looks like it would be confusing to update without context.  I'm guessing this is pulled out so that the descendants of `CaseListMixin` don't inherit this behavior automatically.  If that's the case, I don't think that's something we need to worry about.

The behavior only changes for restricted users, right?  In that case it won't matter for regular users whether this code is here or not.  Restricted users won't be able to access descendants of `CaseListMixin` anyways unless those reports are marked as location-safe, in which case they'd want this behavior.

Would it be possible to just include all this in the `CaseListMixin`?

mkangia replied:
This does change in logic and is not an exact over write. This extract is also used later for any place that needs to add this scope for example in Case Reassignment. There are a lot of classes that inherit from CaseListMixin. I am not sure about adding changes to all those classes by updating CaseListMixin, hence extracted this out. This extract can be used for any places we make locations restricted.

esoergel replied:
What are the changes in logic?  The problem with doing it like this is that it's really difficult to tell how the behavior changes between for instance, the Case List Report and the Reassign Cases Report (since they're supposed to be the same).  Is there another way to fork the logic on `can_access_all_locations`?  It's safe to assume that you can change any behavior when `can_access_all_locations == False`, as long as the behavior is consistent when it's `True`.

mkangia replied:
Right.. that sounds so right.. 
The changes are mostly for restricted user access like not allowing filtering by user types, groups etc or adding a scope filter for restricted user. For a user that can access all locations, i think all remains the same plus the locations integration is added.

mkangia committed:
limit users with limited user ids when provided in submit history report for locations

mkangia committed:
Move CaseExportScope under CaseListMixin itself and add documentation for case owners

mkangia commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -111,45 +130,75 @@ def get_special_owner_ids(self, admin, unknown, demo, commtrack):
     @property
     @memoized
     def case_owners(self):
+        """
```
@esoergel @sheelio Added this documentation to bring to notice what case owners does and would do. Mostly is from the existing implementation but just want to flag couple of things.

> for selected reporting group ids, returns user_ids belonging to these groups
       also finds the sharing groups which has any user from the above reporting group

The second half does not sound right since that would involve all submission by a sharing group which the user just belongs to but has not submitted the cases, hence returning more cases

> selected user ids
       also finds the sharing groups which has any user from the above reporting group

Same as above.

> ids and descendants ids of selected locations and assigned locations of users selected

Similar issue with finding locations of selected users returning a larger set of cases.

esoergel replied:
I'll let @sheelio confirm, but I believe that's expected.  If you select a user or group of users (or location, which is in many ways a group of users), this report should show all cases that those users see on their phones.

mkangia commented:
@esoergel All changes moved to CaseListMixin instead of extracting CaseExportScope. Thanks for pointing this out and forcing this down.

mkangia committed:
Avoid inheriting CaseExportScope and rely on CaseListMixin only as the changes in former are now absorbed in latter

mkangia committed:
avoid passing access permission and use couch_user instead

mkangia committed:
avoiding requesting get_all_users_by_domain and fetch all users at once using UserES instead

mkangia committed:
dimagimon complaint to remove unused variable

esoergel commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -68,14 +70,31 @@ def _build_query(self):
-            query = query.owner(self.case_owners)
+            selected_case_owners = self.case_owners
+            if selected_case_owners:
+                query = query.owner(selected_case_owners)
```
When you deselect everything, the current behavior is to not return any results.  This changes it to return all results, unfiltered (except by scope, where applicable).  I don't think it matters a lot which one we do @sheelio do you have any thoughts?

One possible point of confusion is that if you select something like "Unknown Users", and there are no unknown users on the project, `case_owners` will return `[]`, and then the report will confusingly render with all cases.  I could see users thinking that means that their data is messed up.

mkangia replied:
Oh thats not good. Thanks for catching this. Will revert to old behaviour of not returning any result.

esoergel commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -111,45 +130,75 @@ def get_special_owner_ids(self, admin, unknown, demo, commtrack):
-        selected_sharing_group_ids = EMWF.selected_sharing_group_ids(mobile_user_and_group_slugs)
+        # Get ids for each sharing group that contains a user from selected_reporting_group_users OR
+        # a user that was specifically selected
+        if selected_reporting_group_users or selected_user_ids:
```
Good call.  This'll save us from unnecessary queries in many instances.

esoergel reviewed (commented):
Thank you for following up on my feedback, sorry for making this drag for so long.  I had just a one more point, which should be quick to address

mkangia committed:
revert to blank result in case no case owners are selected

sheelio commented:
Agreed that we should not return any results if the chosen filter is empty.

esoergel commented:
Great, this should be good to go, right?  There's a merge conflict, then you can merge if you think it's ready.

mkangia committed:
ensure accessible forms archived

benrudolph commented:
the conflict is most likely with the PR i've made here: https://github.com/dimagi/commcare-hq/pull/14397

may want to go with the changes i made since i've made `limit_user_ids` work with groups as well as locations (up to you on how you want to merge though)

mkangia committed:
merge master

mkangia committed:
show empty response for empty filters echoing the method used to CaseList

Blank response was the existing behaviour and was retained since returning all accessible results was leading to unexpected response for users with unrestricted access

mkangia committed:
Better message on accessing inaccessible forms when managing forms

mkangia commented:
Thank You @benrudolph. Kept your changes and added mine over them.

esoergel commented:
I approve of the conflict resolution and changes.  The build passed and it's mergeable.  I'm holding off on that because the "Don't Merge" flag is still on.  @mkangia feel free to merge if it's ready.

mkangia committed:
check for inaccessible forms only in case of restricted user when archiving forms

mkangia commented:
Sorry for not merging this yet @esoergel . Moving this to QA along with Manage forms and Reassign Cases.

mkangia committed:
Merge master

dimagimon commented on corehq/apps/reports/standard/cases/basic.py:
```diff
@@ -160,13 +223,16 @@ def case_owners(self):
                                                  selected_user_ids))
                                  .get_ids())
 
+        selected_location_ids = EMWF.selected_location_ids(mobile_user_and_group_slugs)
```
F841 local variable 'selected_location_ids' is assigned to but never used

mkangia committed:
remove unnecessary variable assignment

mkangia committed:
merge mk-243862-locations-submit-history

mkangia committed:
remove duplicate update in context

mkangia committed:
remove duplicate for sharing_group_ids due to merge with master and move filter by users under reporting groups to unrestricted user

mkangia committed:
remove ineffective duplicate for including users at accessible locations during merge with master

mkangia committed:
Update documentation of case_owners for more clarity

mkangia committed:
Merge branch 'mk-243862-locations-submit-history' into mk-243862-reassign-cases-manage-forms

mkangia committed:
Add ability for restricted user to assign cases to accessible locations

mkangia committed:
Merge pull request #14382 from dimagi/mk-243862-reassign-cases-manage-forms

Mk 243862 reassign cases manage forms

mkangia commented:
@esoergel Tested this. Good to go now. 
Already merged in the changes for Reassign Cases and Manage Forms in this.

Would be nice to just give one final look, quite some merge conflicts resolved.

esoergel commented on corehq/apps/groups/models.py:
```diff
@@ -222,6 +222,13 @@ def by_user(cls, user_or_user_id, wrap=True, include_names=False):
+        return [
+            location.case_sharing_group_object() for location in
+            SQLLocation.objects.accessible_to_user(domain, user).filter(location_type__shares_cases=True)
+        ]
```
You could also call this method below in `get_case_sharing_groups` to avoid duplication

esoergel replied:
Oh, never mind, you don't have access to the user in that method.  Probably too much of a pain to change.

esoergel commented on corehq/apps/data_interfaces/tasks.py:
```diff
@@ -43,39 +43,14 @@ def bulk_archive_forms(domain, couch_user, uploaded_data):
 
 @task
-def bulk_form_management_async(archive_or_restore, domain, couch_user, form_ids_or_query_string):
-    # form_ids_or_query_string - can either be list of formids or a BulkFormMangementInterface query url
```
I'm really happy to move this out of here.

esoergel commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -509,11 +522,44 @@ def post(self, request, *args, **kwargs):
+        return xforms_user_ids - accessible_user_ids
+
+    def get_xform_ids(self, request):
+        form_ids_or_query_string = self.get_form_ids_or_query_string(request)
```
I just looked up the definition for this:
```python
    def get_form_ids_or_query_string(self, request):
        if 'select_all' in self.request.POST:
            # Altough evaluating form_ids and sending to task would be cleaner,
            # heavier calls should be in in an async task instead
            import urllib
            form_query_string = urllib.unquote(self.request.POST.get('select_all'))
            return form_query_string
        else:
            return self.request.POST.getlist('xform_ids')
```
It forks depending on parameters, then returns one value, forcing you to check the type of the return again here.  It might be more straightforward to just merge `get_form_ids_or_query_string` in to this method and only do one if/else block (so you can drop the `if type()==list` and `isinstance(basestring)`)

esoergel reviewed (approved):
I had one or two stylistic comments, but nothing blocking.

czue commented:
:tada:

sheelio commented:
Woo hoo!

mkangia commented:
Exciting!!
